### PR TITLE
fix(dropdown): fix dropdown voiceover and refactoring code

### PR DESCRIPTION
**Pull request by Minaparnian**
This change is [Reviewable](https://reviewable.io/reviews/qantasairways/runway/168)

### CODE REVIEW

File: src/components/Dropdown/index.js [hunk 1]
```diff
@@ -3,81 +3,69 @@ import * as React from 'react';
...
+    fontStretch: 'normal',
+    lineHeight: 1.56,
+    letterSpacing: 'normal',
+    background: colours.mediumGrey,
+    borderColor: colours.mediumGrey,
+    color: highlighted ? colours.highlights : colours.white,
+    textDecoration: highlighted ? 'underline' : 'none',
+    position: growMenu ? 'static' : 'relative',
+    width: inline ? 'fit-content' : '100%',
+    height,
```
Comment (congcongcong250) [inline src/components/Dropdown/index.js]: Please merge this height with `height` at **line:43**

File: src/components/Dropdown/index.js [hunk 2]
```diff
@@ -3,81 +3,69 @@ import * as React from 'react';
...
-    maxWidth: '100%',
-    outline: isFocused && !isOpen ? '-webkit-focus-ring-color auto 5px' : 'none'
+    maxWidth: '100%'
   });
 }
 
-export function inputWrapperStyles() {
+export function buttonWrapperStyles({ highlighted, growMenu, inline, height }) {
   return css({
     label: 'runway-dropdown__input-wrapper',
```
Comment (congcongcong250) [inline src/components/Dropdown/index.js]: Please update label text

File: src/components/Dropdown/index.js [hunk 3]
```diff
@@ -3,81 +3,69 @@ import * as React from 'react';
...
-export function dropdownStyles(
-  { highlighted, growMenu, inline, height },
-  isFocused,
-  isOpen
-) {
+export function dropdownStyles({ highlighted, growMenu, inline, height }) {
   return css({
     label: 'runway-dropdown',
-    fontFamily: 'Ciutadella',
     fontSize: '18px',
```
Comment (congcongcong250) [inline src/components/Dropdown/index.js]: Why do we have duplicated style for `dropdown` and `dropdown label`?
I believe we only need once `position: growMenu ? 'static' : 'relative',` as the single source of true for menu positioning.

File: src/components/Dropdown/index.js [hunk 4]
```diff
@@ -107,147 +95,102 @@ export function itemSvgStyles() {
...
-                openMenu();
-              }
-            },
-            onFocus: () => {
-              setFocus(true);
-            }
-          })}
-        />
         <span css={inputSvgStyles(props)}>
           <ChevronDown css={inputSvgStyles(props)} />
```
Comment (congcongcong250) [inline src/components/Dropdown/index.js]: Can we sort out the duplicated style and change the function name to `buttonSvgStyle` ? 
Or a more generic one,`triggerSvgStyle`, so that we don't need to care about the implementation.

File: src/components/Dropdown/index.js [hunk 5]
```diff
@@ -286,60 +229,35 @@ Render.defaultProps = {
...
-        this.validItem(this.props.defaultItemWhenNoneSelected) &&
-        this.itemsContains(this.props.defaultItemWhenNoneSelected)
-      )
-        return this.downshiftSelectItem(this.props.defaultItemWhenNoneSelected);
-      return this.downshiftSelectItem(this.props.items[0]);
+  componentDidUpdate(prevProps) {
+    if (prevProps.initialSelectedItem !== this.props.initialSelectedItem) {
+      /* eslint-disable no-console */
+      console.warn(
+        'Runway Dropdown: initialSelectedItem should not change, please use selectedItem if you want a controlled component'
```
Comment (congcongcong250) [inline src/components/Dropdown/index.js]: Noice!! 👍

File: src/components/Dropdown/index.js [hunk 6]
```diff
@@ -107,147 +97,111 @@ export function itemSvgStyles() {
...
-                setFocus(false);
-                openMenu();
-              }
-            },
-            onFocus: () => {
-              setFocus(true);
-            }
+    return (
+      <div css={{ width: '100%', height: '100%' }}>
+        <button
```
[the remark this answers is not in the document]
Comment (Minaparnian) [reply src/components/Dropdown/index.js]: I feel like to much stuff to class down but I can change the name if you have the better option

File: src/components/Dropdown/index.js (hunk 2 above)
Comment (Minaparnian) [reply src/components/Dropdown/index.js]: Done

File: src/components/Dropdown/index.js (hunk 1 above)
Comment (Minaparnian) [reply src/components/Dropdown/index.js]: what do you mean?

File: src/components/Dropdown/index.js (hunk 4 above)
Comment (Minaparnian) [reply src/components/Dropdown/index.js]: Done

File: src/components/Dropdown/index.js (hunk 1 above)
Comment (Minaparnian) [reply src/components/Dropdown/index.js]: I understand we need to keep both

**Comment by tysonwolker** [comment]
Just add the SelectOnKeyPressContainer.js back and use it as a HoC and I'm happy with this. A lot that can be cleaned up with theming and possibly flex boxing constraints.

File: src/components/Dropdown/index.js (hunk 3 above)
Comment (Minaparnian) [reply src/components/Dropdown/index.js]: Done

File: src/components/Dropdown/index.js (hunk 1 above)
Comment (congcongcong250) [reply src/components/Dropdown/index.js]: There was two duplicate `height` in the previous commit but I believe it's ok now.

File: src/components/Dropdown/index.js [hunk 7]
```diff
@@ -350,27 +263,22 @@ export default class Dropdown extends React.Component {
           }
         }}
         initialSelectedItem={this.getInitialSelectedItem()}
+        selectedItem={this.getSelectedItem()}
```
Comment (congcongcong250) [inline src/components/Dropdown/index.js]: I believe `initialSelectedItem` and `selectedItem` should not be used together.
1. `initialSelectedItem` means Downshift is a **unncontrolled component**,  `selectedItem` as a **control props** instead  is used to make it a **controlled component**
2. If we want to give dev the freedom to choose using Dropdown as controlled or uncontrolled, I reckon to have a condition to check if it's controlled or not before, and passing only one of these two props.

A good example is here: https://github.com/qantasairways/runway/blob/master/src/components/Toggle/index.js#L49

Controlled component:
https://reactjs.org/docs/forms.html#controlled-components
Uncontrolled component:
https://reactjs.org/docs/uncontrolled-components.html

Comment (jamesb3ll) [reply src/components/Dropdown/index.js]: We've added a warning for changing the value of `initialSelectedItem` as previously this was how the component was being updated/controlled, leading to some hacks in componentDidUpdate to maintain the state as a controlled component. Having both `initialSelectedItem` and `selectedItem` above allows the component to be controlled or uncontrolled (which will also be validated items). If they are not provided by the parent, they are not used in downshift. A warning can be added for when using both props together, but I think this should be the job of downshift and not this library, but we can include that in componentDidMount and componentDidUpdate.

Comment (tysonwolker) [reply src/components/Dropdown/index.js]: I think having mixed is fine, most libraries allow both but yeah using one or the other is preferred but doesn't break behaviour however it does indicate poor quality code maybe?

Comment (tysonwolker) [reply src/components/Dropdown/index.js]: Cause going completely uncontrolled is nice but has drawbacks for edgecases which we want to avoid for something like runway.

Comment (congcongcong250) [reply src/components/Dropdown/index.js]: @jamesb3ll the `undefined` is currently handled in all `selectedItem || selectedItem.name` logic.
Warning is a definitely good idea.
I would like to keep `initialSelectedItem ` and `selectedItem` both, as a mixed component with sth like
```js
class DropDown{
    isControlled = () => !!this.props.selectedValue;
    
    render() {
        const valueProps = this.isControlled() ? 
            { selectedItem: this.props.selectedItem } :
            { initialSelectedItem: this.props. initialSelectedItem };

        return (<DownShift {...valueProps}>)
    }
}
```
So that we don't need to care about how Downshift handle the ambiguous scenario that both props are passed down.

Comment (jamesb3ll) [reply src/components/Dropdown/index.js]: The `selectedItem && selectedItem.name` cases are only for if `<Dropdown items={[]} />` is rendered with no other props, I'm not sure what is expected to render here, but we basically render an empty button and empty dropdown (same as previous snapshot before refactor).
Apart from this case, `selectedItem` will never be `undefined` in `downshiftProps`.
When `initialSelectedItem` is provided, downshift will set it's internal state to have `{ selectedItem: this.props.initialSelectedItem }`, so it is always there.
Basically if `selectedItem` is `undefined` and `initialSelectedItem` is a valid item then:
```
this.props. initialSelectedItem === this.props.downshiftProps.selectedItem
```
